### PR TITLE
feat: external engine support 

### PR DESCRIPTION
**Pull request by itaykinnrot**
### Description of the Changes

abstract the engine to use an engine provider and register engines and manage them through it.
We still need to manage the currently selected engine in the provider, like we do for adapters - will be handled in separate PR.

### CheckLists

- [X] changes have been done against master branch, and PR does not conflict
- [X] new unit / functional tests have been added (whenever applicable)
- [X] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/engines/engine-provider.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+   * Add an engine to the registry.
+   * @function register
+   * @param {IEngine} engine -  The engine to register.
+   * @static
+   * @returns {void}
+   */
+  static register(engine: typeof IEngine): void {
+    if (engine) {
+      if (!EngineProvider._engineProviders.includes(engine)) {
+        EngineProvider._logger.debug(`Adapter <${engine.id}> has been registered successfully`);
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: Adapter -> Engine

File: src/engines/engine-provider.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        EngineProvider._engineProviders.push(engine);
+      } else {
+        EngineProvider._logger.debug(`Adapter <${engine.id}> is already registered, do not register again`);
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: Adapter -> Engine

File: src/engines/engine-provider.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    let index = EngineProvider._engineProviders.indexOf(engine);
+    if (index > -1) {
+      EngineProvider._logger.debug(`Unregistered <${engine.id}> adapter`);
+      EngineProvider._engineProviders.splice(index, 1);
+    }
+  }
+
+  /**
+   * Get the appropriate Engines.
+   * @function getMediaSourceAdapter
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: getEngines

File: src/engines/engine-provider.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   * @returns {Array<IEngine>|null} - The Array of engines, or null if such doesn't exists.
+   * @static
+   */
+  static getEngines(): ?Array<typeof IEngine> {
+    if (EngineProvider._engineProviders) {
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: this will always be true

File: src/player.js [hunk 5]
```diff
@@ -157,6 +158,7 @@ export default class Player extends FakeEventTarget {
    */
   static runCapabilities(): void {
     Player._logger.debug("Running player capabilities");
+    Player._engines = EngineProvider.getEngines();
```
Comment (OrenMe) [inline src/player.js]: You should init this in the` _engines` static initializer, and remove the importing of `html5` from here(both require and setting it on the `_engines` array

File: src/playkit.js [hunk 6]
```diff
@@ -72,7 +73,8 @@ export {BaseDrmProtocol};
 // Export the player capabilities
 const getCapabilities = Player.getCapabilities;
 const setCapabilities = Player.setCapabilities;
-export {getCapabilities, setCapabilities};
+
+export {getCapabilities, setCapabilities,registerEngine};
```
Comment (OrenMe) [inline src/playkit.js]: add whitespace indent after comma

File: test/src/engines/engine-provider.spec.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  });
+
+  it('should register Adapter1', () => {
+    EngineProvider._engineProviders.length.should.equal(0);
+    EngineProvider.register(FakeFLASHEngine);
+    EngineProvider._engineProviders.length.should.equal(1);
+    EngineProvider._engineProviders[0].id.should.equal("Flash");
+  });
+
+  it('should not register Adapter1 twice', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `register engine`

File: test/src/engines/engine-provider.spec.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+  beforeEach(() => {
+    EngineProvider._engineProviders = [];
+  });
+
+  after(() => {
+    EngineProvider._engineProviders  = oldEngineProviders;
+  });
+
+  it('should register Adapter1', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `register engine`

File: test/src/engines/engine-provider.spec.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  it('should not register Adapter1 twice', () => {
+    EngineProvider._engineProviders.length.should.equal(0);
+    EngineProvider.register(FakeHTML5Engine);
+    EngineProvider.register(FakeHTML5Engine);
+    EngineProvider._engineProviders.length.should.equal(1);
+    EngineProvider._engineProviders[0].id.should.equal("Html5");
+  });
+
+  it('should register Adapter1 and Adapter2', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `register engine`

File: test/src/engines/engine-provider.spec.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+describe('EngineProvider:unRegister', () => {
+
+  beforeEach(() => {
+    EngineProvider._engineProviders = [FakeHTML5Engine, FakeFLASHEngine];
+  });
+  after(() => {
+    EngineProvider._engineProviders = oldEngineProviders;
+  });
+
+  it('should unRegister Adapter1', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `runegister engine`

File: test/src/engines/engine-provider.spec.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    EngineProvider._engineProviders.length.should.equal(2);
+    EngineProvider.unRegister(FakeHTML5Engine);
+    EngineProvider._engineProviders.length.should.equal(1);
+    EngineProvider._engineProviders[0].id.should.equal("Flash");
+  });
+
+  it('should unRegister Adapter1 and Adapter2', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `unregister engine`

File: test/src/engines/engine-provider.spec.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    EngineProvider._engineProviders.length.should.equal(2);
+    EngineProvider.unRegister(FakeHTML5Engine);
+    EngineProvider.unRegister(FakeFLASHEngine);
+    EngineProvider._engineProviders.length.should.equal(0);
+  });
+
+  it('should do nothing for adapter 3', () => {
```
Comment (OrenMe) [inline test/src/engines/engine-provider.spec.js]: `engine`

File: src/engines/engine-provider.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
+//@flow
+import getLogger from '../utils/logger'
+
+/**
+ * Engine Provider
+ * @classdesc
+ */
+export default class EngineProvider {
```
Comment (dan-ziv) [inline src/engines/engine-provider.js]: Don't export default (we want to align this convention soon)
```
class EngineProvider  { ... }
export {EngineProvider};
```

File: src/engines/html5/html5.js [hunk 14]
```diff
@@ -936,3 +937,5 @@ export default class Html5 extends FakeEventTarget implements IEngine {
     return null;
   }
 }
+
+registerEngine(Html5);
```
Comment (dan-ziv) [inline src/engines/html5/html5.js]: Check if html5 is supported before register it
`if (Html5.isSupported())`

File: src/player.js (hunk 5 above)
Comment (OrenMe) [reply src/player.js]: rethinking this, maybe don't save the `_engines` in here at all and always call `enginesProvider.getEngines` when wanting to access an engine, also have the current selected `_engine' managed by it so you can have `enginesProvider._selectedEngine` saved on the `enginesProvider` and then you can access it via `enginesProvider.getEngine` etc...

File: src/engines/html5/html5.js (hunk 14 above)
Comment (itaykinnrot) [reply src/engines/html5/html5.js]: Today our capabilities framework already include isSupported if we wish to change it  - lets add a task for it :-)

File: src/player.js (hunk 5 above)
Comment (itaykinnrot) [reply src/player.js]: i'll remove the _engines from the player - as for the rest of the logic - i'll open a ticket for it for next version -  i also agree with it.

File: src/engines/html5/html5.js (hunk 14 above)
Comment (OrenMe) [reply src/engines/html5/html5.js]: I think what @dan-ziv means is to not add HTML5 as an option if it's not supported.
Today, as we didn't add any support to platform that didn't support HTML5/MSE playback we didn't care about this.

File: src/engines/engine-provider.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * The logger of the Engine provider.
+   * @member {any} _logger
+   * @static
+   * @private
+   */
+  static _logger: any = getLogger('EngineProvider');
+
+  /**
+   * The Engine adapter registry.
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: remove adapter, just _The engines registry_

File: src/engines/engine-provider.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Get the appropriate Engines.
+   * @function getEngines
+   * @returns {Array<IEngine>|null} - The Array of engines, or null if such doesn't exists.
+   * @static
+   */
+  static getEngines(): ?Array<typeof IEngine> {
+    if (EngineProvider._engineProviders.length > 0) {
```
Comment (OrenMe) [inline src/engines/engine-provider.js]: why do we need this condition at all? the `EngineProvider._engineProviders` is initialized as an empty array.

File: src/playkit.js [hunk 17]
```diff
@@ -69,10 +71,14 @@ export {Env};
 // Export base DRM protocol
 export {BaseDrmProtocol};
 
+//Export default html5 engine
+export {Html5}
```
Comment (OrenMe) [inline src/playkit.js]: why are you exposing this here like this?

File: src/player.js [hunk 18]
```diff
@@ -170,7 +163,7 @@ export default class Player extends FakeEventTarget {
   static getCapabilities(engineType: ?string): Promise<{ [name: string]: any }> {
     Player._logger.debug("Get player capabilities", engineType);
     const promises = [];
```
Comment (odedhutzler) [inline src/player.js]: maybe a more meaningful name, like avaiableEngines (though its not exactly that)

File: src/playkit.js (hunk 17 above)
Comment (itaykinnrot) [reply src/playkit.js]: adding it - so it will be part of the player - and if someone will want to inherit from it and do his own changes

**Pull request merged**